enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+internal loadbalancer implementation has been around for a number of
+releases at this point and we are just extending it.
+
+#### Dev Preview -> Tech Preview
+
+NA
+
+#### Tech Preview -> GA
+
+NA
+
+#### Removing a deprecated feature
+
+NA
+
+### Upgrade / Downgrade Strategy
+
+Upgrades and downgrades will be handled the same way they are for the current
+internal loadbalancer implementation. On upgrade, existing VIP configurations
+will be maintained. We will not (and cannot) automatically add new VIPs to a
+cluster on upgrade. If a deployer of an older dual stack cluster wants the new
+VIP functionality that will have to be a separate operation from upgrade.
+
+### Version Skew Strategy
+
+This will also be handled as it is today. The keepalived configuration used
+for the new version must work with the old version until all nodes have been
+upgraded. If breaking changes are needed, there will be a migration mechanism
+that runs after upgrade is complete (similar to what we did in the multicast
+to unicast migration).
+
+### Operational Aspects of API Extensions
+
+NA
+
+#### Failure Modes
+
+NA
+
+#### Support Procedures
+
+NA
+
+## Implementation History
+
+4.11: Initial implementation
+
+## Drawbacks
+
+Nothing significant. A very small amount of compute resources will be used
+to manage the new VIPs. Even this can be avoided by simply not specifying
+the second VIP if it is not needed.
+
+## Alternatives
+
+Implement this somewhat like the
+[existing workaround](https://github.com/yboaron/kepalived-v6) - create a
+separate instance of keepalived that is responsible for ipv6 only. This
+results in some duplication of containers and configs, but it would probably
+simplify the code changes.
+
+### Alternative install-config layouts
+
+#### Just add v6 versions of the VIP fields
+```yaml
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIP: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is a minimal change, but it leaves us with somewhat of a problem in that
+we can't assume the old fields always contain a v4 address since previous
+releases allowed either v4 or v6. We may be able to deal with that by
+detecting the appropriate ip version for the old fields and populating new
+fields on the api side. E.g. if apiVIP is v4, we populate
+`platformStatus.apiServerInternalIPv4` and if it is v6 we populate
+`apiServerInternalIPv6` instead. That way the only place we need to worry about
+the difference is in the installer. We would need a validation to ensure a v6
+VIP isn't specified in both fields.
+
+Note that this assumes we replace the existing api fields with two new ones
+per VIP and do not reuse the existing ones at all. I think we're going to want
+to do that regardless of what install-config layout we use so we don't need
+logic on the backend in machine-config-operator and baremetal-runtimecfg to
+handle older cluster configs.
```

Comment:
There should not be any logic needed in machine-config-operator or baremetal-runtimecfg to handle older cluster configs. The installer would need to populate the new fields with the values from the deprecated fields. The backend should be able to rely on the new fields always being used.

Reply:
MCO doesn't parse the install config for day 2 operations.

Reply:
This might be a terminology problem. When I say "cluster config", I'm referring to the Infrastructure and ControllerConfig objects that are populated with the values specified in install-config. I will be more explicit in the next revision.

The main thing is that if we deprecate a field in install-config, that will require deprecation of an API field and migration of the values on the backend as well. We can do that, but it is a small drawback to some of the options.

Reply:
OK. It was confusing to include cluster config then in the discussion of the merits of alternative designs for the install config. I view the layout of the install-config fields and the layout of the config.openshift.io API fields as separate concerns. If there is nothing using the install-config in-cluster, then the backend arguments has even less bearing as criteria for which install-config alternative to select.